PFE Registry Secret API /api/v1/registrysecrets

maysunfaisal opened the pull request:
This is for Issue #665 

This PR 
- introduced APIs as per comment https://github.com/eclipse/codewind/issues/665#issuecomment-550557395
- Tested to make sure the APIs work(see screenshot below) and also tested to make sure latest works with the APIs checked in

Tested the APIs with two sets of docker registries:
![Screen Shot 2019-11-15 at 2 02 37 PM](https://user-images.githubusercontent.com/31771087/68974130-f9d63580-07bd-11ea-99a0-31953c47dca7.png)
![Screen Shot 2019-11-15 at 2 05 44 PM](https://user-images.githubusercontent.com/31771087/68974131-f9d63580-07bd-11ea-89f8-e1471e8bfda2.png)
![Screen Shot 2019-11-15 at 2 06 16 PM](https://user-images.githubusercontent.com/31771087/68974132-fa6ecc00-07bd-11ea-8605-aad7fd35a41d.png)
![Screen Shot 2019-11-15 at 2 07 51 PM](https://user-images.githubusercontent.com/31771087/68974133-fa6ecc00-07bd-11ea-9964-e5573684a453.png)

P.S. The codewind-che-plugin PR https://github.com/eclipse/codewind-che-plugin/pull/106 also depends on this PR.

maysunfaisal commented:
@rajivnathan @johnmcollier can you pls review from Turbine's perspective?

@hhellyer @stalleyj can you pls review from Portal's perspective?

tetchel commented:
portal's TRACE logging which logs all request bodies is going to log user passwords sent to this endpoint, we are going to have to fix that

maysunfaisal commented:
@tetchel you're right, good catch

```
[15/11/19 20:54:27 server.js] [TRACE] Checking origin host  matches expected host localhost:9191
[15/11/19 20:54:27 server.js] [TRACE] POST /api/v1/registrysecrets called with 
 req.body { url: 'docker.io',
  username: 'maysunfaisal',
  password: 'yourenotsupposedtoseeme' }
[15/11/19 20:54:27 /portal/routes/registrySecrets.route.js] [DEBUG] POST /api/v1/registrysecrets called
```

@stalleyj @hhellyer any idea how I would disable that?

tetchel commented on src/pfe/portal/routes/registrySecrets.route.js:
```diff
@@ -0,0 +1,85 @@
+    let user = req.cw_user;
+    log.debug(`POST /api/v1/registrysecrets called`);
+    const username = req.sanitizeBody('username');
+    const password = req.sanitizeBody('password');
```
good catch, we could encode the password to get around this

maysunfaisal replied:
so after all we'll be sending encoded password. Then do we go with what I suggested earlier:

encoding

```
credentials[registryUrl] = {
      username: userName,
      password: userPassword
    };
``` 

and sending it as a single property in the req body? or just encode the password?

tetchel replied:
it makes sense to do something like

```
> const credentials = { username: "timetchells", password: "pw123" }
undefined
> const credentialsStr = JSON.stringify(credentials);
undefined
> const encodedCreds = Buffer.from(credentialsStr).toString("base64")
undefined
> encodedCreds
'eyJ1c2VybmFtZSI6InRpbWV0Y2hlbGxzIiwicGFzc3dvcmQiOiJwdzEyMyJ9'
```

then in the request body we send `{ credentials: encodedCreds }` and you can decode and then parse it back to an object.

maysunfaisal replied:
updated

maysunfaisal commented on src/pfe/portal/modules/User.js:
```diff
@@ -53,6 +58,12 @@ module.exports = class User {
     }
     this.secure = true;
+    this.dockerConfigFile = "/root/.docker/config.json";
+    this.codewindPFESecretName = "codewind-pfe-docker-registries";
```
workspaceID will not work in Hybrid, since there may not be a Che and we wont have the workspaceID.

johnmcollier replied:
@maysunfaisal Portal & the Cwctl folks are still using a workspaceID, they're generating it from the timestamp when PFE is deployed

maysunfaisal replied:
So are we still going to have it in the PFE ENV VAR even on Hybrid? If it is, I will make the change for all review comments

johnmcollier replied:
They're actually still using `CHE_WORKSPACE_ID` https://github.com/eclipse/codewind-installer/blob/master/pkg/remote/deploy_pfe.go#L85

That should probably be renamed on both ends to be `WORKSPACE_ID`

maysunfaisal replied:
ok cool, i will make the change with what we currently have

maysunfaisal replied:
updatted

maysunfaisal commented on src/pfe/portal/modules/User.js:
```diff
@@ -659,6 +670,288 @@ module.exports = class User {
+        }
+
+        // Get the Codewind PFE deployment name and uid labeled with app=codewind-pfe
+        resp = await this.k8client.apis.apps.v1.namespaces(process.env.KUBE_NAMESPACE).deployments.get({ qs: { labelSelector: "app=codewind-pfe" } });
```
WorkspaceID may not be there on hybrid. What else can be unique? Thats why i didnt go with it

do we want to deploy PFE on hybrid with a unique workspace id as well?

johnmcollier replied:
See my other comment - workspaceID will still be there on Hybrid https://github.com/eclipse/codewind-installer/blob/master/pkg/remote/deploy_pfe.go#L85

maysunfaisal replied:
updated

maysunfaisal commented on src/pfe/portal/modules/User.js:
```diff
@@ -53,6 +58,12 @@ module.exports = class User {
+    this.codewindPFESecretName = "codewind-pfe-docker-registries";
+    // Get the Kube Client context when running in K8s
+    if (global.codewind.RUNNING_IN_K8S == true) {
+      this.k8client = new K8Client({ config: k8config.getInCluster(), version: '1.9' });
```
Not sure about this. Will need to check. Portal is already doing it this way in their server.js

maysunfaisal replied:
So I had initially had the Kube client initialized in a single place(ie Portal and Portal was telling Turbine during PFE startup about the Kube Client obj) across all the Codewind codebase but after thinking about it, i think it is cleaner and better if Portal & Turbine had their own Kube Client initialization since they're different modules.

So I have updated the Portal code and I am only using the initialized Kube Client and not doing any re-initialization. The change is in the commit https://github.com/eclipse/codewind/pull/1118/commits/0ebb283e946cb4f21fe18051068a806024e9be30

maysunfaisal commented on src/pfe/portal/modules/User.js:
```diff
@@ -659,6 +670,288 @@ module.exports = class User {
+   * Function to setup the Docker config file and Kube secret
+   */
+  async setupRegistrySecret(username, password, url) {
+    log.info("Setting up the Docker Registry secret for url: " + url + " and username: " + username);
```
Does it matter? Because username is public to everyone who has access to CW. But if we are worried about it being retinaed in the logs, we can remove it

maysunfaisal commented on src/pfe/portal/routes/registrySecrets.route.js:
```diff
@@ -0,0 +1,85 @@
+    const password = req.sanitizeBody('password');
+    const url = req.sanitizeBody('url');
+
+    if (!username || !password || !url) {
```
sure, I can look at what the other APIs are doing

maysunfaisal replied:
updated

maysunfaisal commented on src/pfe/portal/modules/utils/Logger.js:
```diff
@@ -71,8 +71,15 @@ class Logger {
+    }
     let msg = `${route} called`;
-    if (!this.log.isTraceEnabled()) {
+    if (!this.log.isTraceEnabled() || isRegistrySecretRoute) {
```
good point, and this sounds much safer. I shall make the change 👍

maysunfaisal replied:
updated

johnmcollier commented on src/pfe/file-watcher/scripts/root-watcher.sh:
```diff
@@ -91,6 +95,7 @@ if [ "$IN_K8" == "true" ]; then
 	echo "Initializing a custom helm tiller"
 	helm init --upgrade --service-account $SERVICE_ACCOUNT_NAME
 
+	# TO BE REMOVED AS PART OF CODEWIND ISSUE https://github.com/eclipse/codewind/issues/665 WHEN APIS ARE READY TO BE CONSUMED
```
What happens when PFE restarts and the .dockerconfigjson under `/root/.docker/` is lost?

I'm wondering if we instead want to adapt this for new architecture and have PFE pull down the config.json from the secret when it starts up, if a secret already exists.

maysunfaisal replied:
so we want to do this in root-watcher.sh?

johnmcollier replied:
We could adapt the existing code to pull down the .dockerconfigjson with kubectl _or_ handle it elsewhere. It doesn't matter where specifically, so long as the .dockerconfigjson isn't lost when PFE restarts

maysunfaisal replied:
So the Q is how does the logic decide if it was a PFE restart? Since when PFE starts there never is a `/roott/.docker/config.json` until the API is called

johnmcollier replied:
Well, in `root-watcher.sh` (the closest to a Turbine entrypoint there is), we could check for a registry secret labeled with the instance's workspace ID. If it exists, we pull down the `.dockerconfigjson` file under `/root/.docker/`. If no secret exists, we create an empty `/root/.docker`

maysunfaisal replied:
yeah makes sense, tho its gonna involve using kubectl and base64/openssl commands

maysunfaisal replied:
updated behind a flag for safe keeping

rajivnathan commented on src/pfe/portal/modules/User.js:
```diff
@@ -659,6 +670,288 @@ module.exports = class User {
+
+    // Only update the Kube secret when running in K8s
+    let isServiceAccountPatched = false;
+    if (createKubeSecret && global.codewind.RUNNING_IN_K8S) {
```
As discussed should check what we want to do for the non-K8s scenario and test it out to make sure the response is what we'd expect.

maysunfaisal replied:
updated

maysunfaisal commented on docs/openapi.yml:
```diff
@@ -1111,6 +1111,140 @@ paths:
+        400:
+          description: Bad Request
+          content:
+            text/plain:
```
nope, Time wanted it this way

maysunfaisal commented on src/pfe/portal/modules/User.js:
```diff
@@ -659,6 +670,288 @@ module.exports = class User {
+        const encodedDockerConfig = btoa(JSON.stringify(jsonObj));
+
+        // Get the Kube Secret and delete if it exists
+        let resp = await this.k8client.api.v1.namespaces(process.env.KUBE_NAMESPACE).secret.get({ qs: { labelSelector: "app=codewind-pfe" } });
```
updated

maysunfaisal commented on src/pfe/portal/modules/User.js:
```diff
@@ -659,6 +670,288 @@ module.exports = class User {
+          "kind": "Secret",
+          "metadata": {
+            "labels": {
+              "app": "codewind-pfe"
```
updated

maysunfaisal commented:
@hhellyer Can you pls review? I have made the changes you requested and I have also updated Portal `server.js` a bit to re-use the initialized Kube Client obj in `User.js`

@johnmcollier @rajivnathan can you pls review the updates? I have addressed all the comments.

hhellyer commented on src/pfe/portal/modules/User.js:
```diff
@@ -34,13 +34,13 @@ const util = require('util');
 
-  static async createUser(user_id, userString, workspace, uiSocket) {
-    let user = new User(user_id, userString, workspace, uiSocket);
+  static async createUser(user_id, userString, workspace, uiSocket, k8Client) {
```
An alternative to this could be adding `global.codewind.K8S_CLIENT`, I'm not totally happy with passing the client as another argument since it would be the same client for all users (if we had more than one) but I'm also not that happy about adding another global. :-)

maysunfaisal replied:
updated to incl in global obj

maysunfaisal replied:
gonna push the changes in a new PR

maysunfaisal commented on src/pfe/portal/modules/User.js:
```diff
@@ -659,6 +665,312 @@ module.exports = class User {
+            }
+          ]
+        };
+        resp = await await this.k8Client.api.v1.namespaces(process.env.KUBE_NAMESPACE).serviceaccounts(process.env.SERVICE_ACCOUNT_NAME).patch({body: patch});
```
updated

maysunfaisal replied:
gonna push the changes in a new PR

maysunfaisal commented on src/pfe/portal/modules/User.js:
```diff
@@ -659,6 +665,312 @@ module.exports = class User {
+    }
+
+    // Only update the Kube secret when running in K8s
+    let isServiceAccountPatched = false;
```
updated, instead of assuming as true

I updated the if block - `if (isServiceAccountPatched || !global.codewind.RUNNING_IN_K8S) {`

maysunfaisal replied:
gonna push the changes in a new PR

maysunfaisal commented on src/pfe/portal/modules/User.js:
```diff
@@ -659,6 +665,312 @@ module.exports = class User {
+      const retval = {
+        statusCode: 500,
+        body: msg
+      };
```
gonna push the changes in a new PR

maysunfaisal commented:
@rajivnathan @johnmcollier @tetchel can any one of you pls help merge? thx
